own: add new `select:file.owners` search job

leonore opened the pull request:
closes #46990 

- add ability for `select:file.owners` search jobs 
- owners are resolved using the codeowners file information and the users and teams tables
- the search stream API has new responses for different owner types (person, team, unknown)

this is guarded under feature flag `search-ownership`
## Test plan

Added unit tests for `ResolveOwnersWithType` and `SelectOwnersJob`

Manually tested with stream API.
```
curl \
  -H 'Authorization: token x' -H 'Accept: text/event-stream' --get --url 'https://sourcegraph.test:3443/.api/search/stream' --data-urlencode "q=repo:sourcegraph-test leo select:file.owners"

# ✨✨✨✨✨✨✨✨✨✨ this is where the magic happens ✨✨✨✨✨✨✨✨✨✨✨✨✨✨
event: matches
data: [{"type":"person","handle":"","email":"","username":"notadmin","displayName":"","avatarURL":""},{"type":"person","handle":"","email":"","username":"notadmin","displayName":"","avatarURL":""},{"type":"unknownOwner","handle":"leonore","email":""}]
# ✨✨✨✨✨✨✨✨✨✨ this is where the magic happens ✨✨✨✨✨✨✨✨✨✨✨✨✨✨

event: progress
data: {"done":true,"repositoriesCount":1,"matchCount":3,"durationMs":88,"skipped":[]}

event: done
data: {}
```

camdencheek commented on internal/search/codeownership/select_job.go:
```diff
@@ -0,0 +1,107 @@
+		event.Results, err = getCodeOwnersFromMatches(ctx, clients.Gitserver, &rules, event.Results)
+		if err != nil {
+			mu.Lock()
+			errs = errors.Append(errs, err)
```
Do you still want to send the event even if there is an error?

camdencheek commented on internal/search/codeownership/select_job.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func (s *selectOwnersJob) Fields(v job.Verbosity) (res []otlog.Field) {
+	return res
```
minor: the `owners` field should be returned here for debugging/tracing

camdencheek commented on internal/search/result/owner.go:
```diff
@@ -0,0 +1,72 @@
+	"github.com/sourcegraph/sourcegraph/internal/types"
+)
+
+type OwnerMatch struct {
+	*codeownerspb.Owner
+
+	// The following contain information about what search the owner was matched from.
+	InputRev *string           `json:"-"`
+	Repo     types.MinimalRepo `json:"-"`
+	CommitID api.CommitID      `json:"-"`
+	Path     string
+
+	LimitHit int
+
+	// Debug is optionally set with a debug message explaining the result.
+	//
+	// Note: this is a pointer since usually this is unset. Pointer is 8 bytes
+	// vs an empty string which is 16 bytes.
+	Debug *string `json:"-"`
+}
```
Question here: do we want the API object to an "owner" or do we want it to be a "person"/"team"? I imagine it's more generally useful to have a "person" type and a "team" type rather than the very specific "owner" type.

limitedmage replied:
+1, I'd rather this be "person"/"team" than "owner"; we can return different data and render both in different ways in the UI as well (eg. show team members)

cbart replied:
That's the idea.

The way I see it, with the data that comes through the teams/members API, we would be able to discern handles to be either teams or people. But this won't always be the case I think. We need to also have a good way of representing an ownership handle without knowing whether it's a team or a person.

camdencheek commented on internal/search/result/owner.go:
```diff
@@ -0,0 +1,72 @@
+type OwnerMatch struct {
+	*codeownerspb.Owner
+
+	// The following contain information about what search the owner was matched from.
+	InputRev *string           `json:"-"`
+	Repo     types.MinimalRepo `json:"-"`
+	CommitID api.CommitID      `json:"-"`
+	Path     string
+
+	LimitHit int
+
+	// Debug is optionally set with a debug message explaining the result.
+	//
+	// Note: this is a pointer since usually this is unset. Pointer is 8 bytes
+	// vs an empty string which is 16 bytes.
+	Debug *string `json:"-"`
```
Once we have an owner, do we care about any of this?

camdencheek commented on internal/search/result/owner.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func (om *OwnerMatch) RepoName() types.MinimalRepo {
+	// todo(own): this might not make sense forever. right now we derive ownership from files within a repo but if we
+	// extend this with external sources then it might not be mandatory to attach an owner to repo.
+	// as an alternative we can also conduct a check that nothing expects RepoName to always exist.
```
Yeah, if we're adding more result types that aren't necessarily tied to repos, it might make sense to remove this from the interface.

camdencheek commented on internal/search/result/owner.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func (om *OwnerMatch) Select(selectPath filter.SelectPath) Match {
+	// todo can this be safely ignored?
```
Yep, you're good here. For now, I don't expect there is anything you can "select" from an owner, so returning `nil` is correct

camdencheek reviewed (commented):
Looking good!

limitedmage commented on internal/search/streaming/http/events.go:
```diff
@@ -167,6 +167,16 @@ type EventCommitMatch struct {
+	// Type is always OwnerMatchType. Included here for marshalling.
+	Type MatchType `json:"type"`
+
+	Handle string `json:"handle"`
```
I'm thinking the handle should be some kind of "fully qualified name" that includes the full URL for disambiguation (eg. `github.com/username` instead of just `username`). If we wanted to match with a Sourcegraph user, we can also return "sourcegraphUsername" or similar.

leonore replied:
the `Handle` here is 1-1 with the `Handle` in the owner proto. this is linked to your comment above, but I need to do some digging to find out how we can augment the user information based on handle/email info from the codeowners file

limitedmage replied:
Yeah I think we need to, both here and in the Owner proto, have the codehost ID/URL of where this handle is from; otherwise I don't see this being very useful, especially in multi-codehost environments.

limitedmage commented on internal/search/streaming/http/events.go:
```diff
@@ -167,6 +167,16 @@ type EventCommitMatch struct {
 
 func (e *EventCommitMatch) eventMatch() {}
 
+type EventOwnerMatch struct {
```
We need to return some kind of URL with this; it could be a Sourcegraph user profile (if we match to an existing user) or an external link to the code host's user profile if we find it.

leonore replied:
are we hoping to be able to match a user profile from a handle (retrieved from the owner proto)? do we have guarantees that the handle would match a user's such that we could use some existing graphql endpoints for this?

limitedmage replied:
We don't have guarantees, but if the user exists with that code host handle or email, we should link to them. If the user doesn't exist, we can link to the user profile on the code host. WDYT?

limitedmage replied:
We would need a way to know what that code host is (eg. from the repo URL and code host type, eg. github.com is a  GitHub type codehost) to create URLs in the right format. If it's an unknown/unspported code host (eg. generic Git repo or Perforce) we can show a warning saying that we found the owner's handle but not sure where it's from.

camdencheek commented on cmd/frontend/backend/own.go:
```diff
@@ -63,3 +79,74 @@ func (s ownService) OwnersFile(ctx context.Context, repoName api.RepoName, commi
+			continue
+		}
+		resolved = append(resolved, resolvedOwner)
+		s.ownerCache[po.Handle+po.Email] = resolvedOwner
```
Can this method be called concurrenty? If so, the cache accesses will be racy.

camdencheek replied:
Nvm, I see that we protect all calls with a mutex. It's a little unintuitive that a "service" cannot be used concurrently though -- I'd recommend adding a mutex to protect the cache to avoid issues in the future.

leonore replied:
good catch. we've been bit by this in the past- when we replicated searcher logic to use in search aggregations and were storing results in memory, and forgot to add [a lock in the caller](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/commit/40068cdd85c5720114d8d7677a0c4236e1a10de7?visible=2). I'll add a lock in the service just to be safe.

camdencheek commented on cmd/frontend/internal/search/search.go:
```diff
@@ -516,6 +519,35 @@ func fromCommit(commit *result.CommitMatch, repoCache map[api.RepoID]*types.Sear
+			Email:  v.Email,
+		}
+	}
+	// We shouldn't reach this.
+	return &streamhttp.EventUnknownOwnerMatch{
+		Type:   streamhttp.UnknownOwnerMatchType,
+		Handle: owner.ResolvedOwner.Identifier(),
+	}
```
If this is unreachable, should we just panic here so the problem is at least visible?

leonore replied:
I'm about to change this so that we return an owner match by default, I just need to change the ResolvedOwner fields again. ty!

camdencheek reviewed (changes requested):
There are a few places left where we need to handle owner match: [1](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/enterprise/internal/compute/output_command.go?L102-103), [2 (optional)](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/enterprise/cmd/frontend/internal/compute/resolvers/resolvers.go?L184-184), [3](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/internal/search/job/jobutil/sub_repo_perms_job.go), [4](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/cmd/frontend/graphqlbackend/search_results.go?L166-166), [5](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/cmd/frontend/graphqlbackend/search_results.go?L332-332)

leonore commented:
> There are a few places left where we need to handle owner match: [1](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/enterprise/internal/compute/output_command.go?L102-103), [2 (optional)](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/enterprise/cmd/frontend/internal/compute/resolvers/resolvers.go?L184-184), [3](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/internal/search/job/jobutil/sub_repo_perms_job.go), [4](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/cmd/frontend/graphqlbackend/search_results.go?L166-166), [5](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c3edd7d3ce12001873abbe963c77392a5e0ed81b/-/blob/cmd/frontend/graphqlbackend/search_results.go?L332-332)

@camdencheek 1 and 2 as referenced are from compute. I've added some basic handling so it doesn't panic but supporting owner in compute should be part of a secondary effort, if any. I'm super unfamiliar with the internals, and I think we've marked it as deprecated?

for 4- do you mind if I open a separate PR for this? as far as I can see this is just the graphql resolvers for the owner result. which is another thing I blissfully hadn't though about since we mostly use the stream API, but I guess we'll still want to support the graphql resolver especially for src-cli. I did think this code was too lightweight for the purpose of adding a new search type 🫠
for 5- I added a skip like for repo based on my quick understanding of the search code there, but I'm not familiar with `sparkline`.

camdencheek commented:
@leonore Oh, for all of these, just "don't panic" is enough for this PR.  I'm sorry I didn't clarify in my comment!

cbart commented on internal/search/codeownership/select_job_test.go:
```diff
@@ -0,0 +1,193 @@
+		autogold.Want("owner matches are as expected", want).Equal(t, matches)
+	})
+	t.Run("deduplicates owner matches", func(t *testing.T) {
+		// See result/owner.go:55 for comment.
```
Is there a way to partially aggregate data like paths (say aggregate them up until it's 5 of them) but otherwise return distinct owners?

cbart replied:
> the only worry I have with this is it would be pretty different from how matches are operated in search now

So perhaps we should not include the paths just yet? Personally I'd rather get de-duplicated list without reasoning, than duplicates with per-file reasoning. @eseliger ideas?

> it would make sub-repo permission filtering harder

Tell me more?

leonore replied:
> So perhaps we should not include the paths just yet?

I think path isn't necessarily expected everywhere, so we could do that. I'll have a closer look tomorrow

> it would make sub-repo permission filtering harder

if you look in this current PR: https://github.com/sourcegraph/sourcegraph/pull/47036/files#diff-ee277a1a00281cb2eaab71f1635abc3e8391f713b762ef5de469e560c7d6554f sub-repo perms filtering is done with path/repo information. so I guess we'd have to iterate over the list of paths attached to the owner and remove any that the actor does not have access to. not impossible but a bit more hacky? ultimately I think we just need to think about whether we want to surface information about _which_ path the owner owns. if the answer is yes, then this is something we'll need to do.

cbart reviewed (approved):
Good PR.

Please feel free to ignore the stylistic comments - just there for discussion.

The only think I'm wondering about is the de-duplication feature you highlighted. Is there a way we could partially aggregate matched paths (say up to 5 or 10)? This way we could testify to the user how the owner was matched, but at the same time we would return an owner only once.

leonore commented on cmd/frontend/backend/own.go:
```diff
@@ -63,3 +82,79 @@ func (s ownService) OwnersFile(ctx context.Context, repoName api.RepoName, commi
+type teamGetterFunc func(context.Context, string) (*types.Team, error)
+
+func tryGetUserThenTeam(ctx context.Context, identifier string, userGetter userGetterFunc, teamGetter teamGetterFunc) (codeowners.ResolvedOwner, error) {
+	user, err := userGetter(ctx, identifier)
+	if err != nil {
+		if errcode.IsNotFound(err) {
+			if teamGetter != nil {
+				team, err := teamGetter(ctx, identifier)
+				if err != nil {
+					return nil, err
+				}
+				return &codeowners.Team{Team: team, OwnerIdentifier: identifier}, nil
+			}
+		}
+		return nil, err
+	}
+	return &codeowners.Person{User: user, OwnerIdentifier: identifier}, nil
```
I'll have a look at your suggestion locally and pick. On a quick glance it looks nice. you should have seen how bad the indentation was without the refactored helpers 😄

leonore replied:
I think for now what I have communicates the logic of the function a bit better, and I'm not sure this actually works with  the `unknownOwnerOrError` function. for now I'll just keep it as is to :smash-it-in:

cbart replied:
👍

camdencheek replied:
I'm disappointed we don't have a :smash-it-in: reaction in slack
